heron/dsl/src/java/com/twitter/heron/dsl/impl/StreamletImpl.java

```diff
@@ -0,0 +1,296 @@
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package com.twitter.heron.dsl.impl;
+
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+
+import com.twitter.heron.api.topology.TopologyBuilder;
+import com.twitter.heron.dsl.KVStreamlet;
+import com.twitter.heron.dsl.KeyValue;
+import com.twitter.heron.dsl.SerializableBiFunction;
+import com.twitter.heron.dsl.SerializableBinaryOperator;
+import com.twitter.heron.dsl.SerializableFunction;
+import com.twitter.heron.dsl.SerializablePredicate;
+import com.twitter.heron.dsl.SerializableSupplier;
+import com.twitter.heron.dsl.Streamlet;
+import com.twitter.heron.dsl.WindowConfig;
+import com.twitter.heron.dsl.WindowInfo;
+import com.twitter.heron.dsl.impl.streamlets.FilterStreamlet;
+import com.twitter.heron.dsl.impl.streamlets.FlatMapStreamlet;
+import com.twitter.heron.dsl.impl.streamlets.KVFlatMapStreamlet;
+import com.twitter.heron.dsl.impl.streamlets.KVMapStreamlet;
+import com.twitter.heron.dsl.impl.streamlets.LogStreamlet;
+import com.twitter.heron.dsl.impl.streamlets.MapStreamlet;
+import com.twitter.heron.dsl.impl.streamlets.ReMapStreamlet;
+import com.twitter.heron.dsl.impl.streamlets.ReduceByWindowStreamlet;
+import com.twitter.heron.dsl.impl.streamlets.SupplierStreamlet;
+import com.twitter.heron.dsl.impl.streamlets.UnionStreamlet;
+
+/**
+ * A Streamlet is a (potentially unbounded) ordered collection of tuples.
+ * Streamlets originate from pub/sub systems(such Pulsar/Kafka), or from
+ * static data(such as csv files, HDFS files), or for that matter any other
+ * source. They are also created by transforming existing Streamlets using
+ * operations such as map/flatMap, etc.
+ * Besides the tuples, a Streamlet has the following properties associated with it
+ * a) name. User assigned or system generated name to refer the streamlet
+ * b) nPartitions. Number of partitions that the streamlet is composed of. Thus the
+ *    ordering of the tuples in a Streamlet is wrt the tuples within a partition.
+ *    This allows the system to distribute  each partition to different nodes across the cluster.
+ * A bunch of transformations can be done on Streamlets(like map/flatMap, etc.). Each
+ * of these transformations operate on every tuple of the Streamlet and produce a new
+ * Streamlet. One can think of a transformation attaching itself to the stream and processing
+ * each tuple as they go by. Thus the parallelism of any operator is implicitly determined
+ * by the number of partitions of the stream that it is operating on. If a particular
+ * tranformation wants to operate at a different parallelism, one can repartition the
+ * Streamlet before doing the transformation.
+ */
+public abstract class StreamletImpl<R> implements Streamlet<R> {
```

Comment:
maybe BaseStreamlet?

Reply:
Which one are you referring to? Streamlet or StreamletImpl?